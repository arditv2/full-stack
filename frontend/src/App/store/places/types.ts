export const GET_PLACES = '@@places/GET_PLACES';
export const GET_PLACES_SUCCESS = '@@places/GET_PLACES_SUCCESS';
export const GET_PLACES_ERROR = '@@places/GET_PLACES_ERROR';

export interface ILocation {
	lat: string;
	lng: string;
}

export interface IPlace {
	id: string;
	provider: string;
	name: string;
	address: string;
	location?: ILocation;
}

export interface IGetPlaces {
	type: typeof GET_PLACES;
}

export interface IGetPlacesSuccess {
	type: typeof GET_PLACES_SUCCESS;
	payload: IPlace[];
}

export interface IGetPlacesError {
	type: typeof GET_PLACES_ERROR;
	payload: string;
}

export type DispatchTypes = IGetPlaces | IGetPlacesSuccess | IGetPlacesError;

export interface IPlacesState {
	loading: boolean;
	error?: string;
	places?: IPlace[];
}

export interface ISearchForm {
	name: string;
	lat: string;
	lng: string;
}

export interface IApiClient {
	get: (query: string) => Promise<any>;
}

export interface IApi {
	getPlaces: (payload: ISearchForm) => Promise<IPlace[]>;
}
