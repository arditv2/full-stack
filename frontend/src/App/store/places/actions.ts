import { ThunkDispatch } from 'redux-thunk';

import {
	GET_PLACES,
	GET_PLACES_SUCCESS,
	GET_PLACES_ERROR,
	DispatchTypes,
	IGetPlaces,
	IGetPlacesSuccess,
	IGetPlacesError,
	IApi,
	ISearchForm,
} from './types';

import Api from './api';

export const makeGetPlacesAction = (Api: IApi) => (
	payload: ISearchForm
) => async (dispatch: ThunkDispatch<any, any, DispatchTypes>) => {
	try {
		dispatch<IGetPlaces>({
			type: GET_PLACES,
		});

		const places = await Api.getPlaces(payload);

		dispatch<IGetPlacesSuccess>({
			type: GET_PLACES_SUCCESS,
			payload: places,
		});
	} catch (error) {
		dispatch<IGetPlacesError>({
			type: GET_PLACES_ERROR,
			payload:
				(error.response &&
					error.response.data &&
					error.response.data.message) ||
				error.message,
		});
	}
};

export const getPlacesAction = makeGetPlacesAction(Api);
