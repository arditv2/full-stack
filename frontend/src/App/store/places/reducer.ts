import { Reducer } from 'redux';
import {
	IPlacesState,
	GET_PLACES,
	GET_PLACES_SUCCESS,
	GET_PLACES_ERROR,
	DispatchTypes,
} from './types';

const initialState: IPlacesState = {
	loading: false,
};

export const placesReducer: Reducer<IPlacesState> = (
	state: IPlacesState = initialState,
	action: DispatchTypes
): IPlacesState => {
	switch (action.type) {
		case GET_PLACES:
			return {
				loading: true,
			};
		case GET_PLACES_SUCCESS:
			return {
				loading: false,
				places: action.payload,
			};
		case GET_PLACES_ERROR:
			return {
				loading: false,
				error: action.payload,
				places: [],
			};
		default:
			return state;
	}
};
