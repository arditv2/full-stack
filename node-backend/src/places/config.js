require('dotenv').config();

module.exports = Object.freeze({
	services: {
		google: {
			basePath: 'https://maps.googleapis.com/maps/api',
			key: process.env.GOOGLE_API_KEY,
		},
	},
});
