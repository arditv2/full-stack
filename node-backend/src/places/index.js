const axios = require('axios');

const makeRequestHandler = require('./request-handler');
const makeService = require('./service');
const config = require('./config');

const service = makeService({ client: axios, config });
const requestHandler = makeRequestHandler({ service });

module.exports = requestHandler;
