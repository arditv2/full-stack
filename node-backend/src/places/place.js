const makePlace = (id, name, address, { lat, lng } = {}, provider) => {
	if (!id) {
		throw new Error('Place must have an Id');
	}
	if (!name) {
		throw new Error('Place must have a name');
	}
	if (!address) {
		throw new Error('Place must have an address');
	}
	if (!provider) throw new Error('Place must have a provider');

	return Object.freeze({
		id,
		name,
		address,
		location: lat &&
			lng && {
				lat,
				lng,
			},
		provider,
	});
};

module.exports = makePlace;
