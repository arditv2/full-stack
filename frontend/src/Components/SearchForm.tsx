import React, { FC, useState } from 'react';
import Form from 'react-bootstrap/Form';
import Button from 'react-bootstrap/Button';
import Col from 'react-bootstrap/Col';

import { ISearchForm } from '../App/store/places/types';

interface IProps {
	handleSubmit: (payload: ISearchForm) => void;
}

const initialState: ISearchForm = {
	name: '',
	lat: '',
	lng: '',
};

const SearchForm: FC<IProps> = ({ handleSubmit }) => {
	const [state, setState] = useState<ISearchForm>(initialState);
	return (
		<Form className='mt-3 mb-3'>
			<Form.Group controlId='formPlaceName'>
				<Form.Label>Place name</Form.Label>
				<Form.Control
					type='text'
					placeholder='Enter place name'
					data-test-id='nameInput'
					value={state.name}
					onChange={(e) =>
						setState({
							...state,
							name: e.target.value,
						})
					}
				/>
			</Form.Group>
			<Form.Row>
				<Form.Group as={Col} md='6' controlId='formLat'>
					<Form.Label>Latitude</Form.Label>
					<Form.Control
						type='text'
						placeholder='Enter latitude'
						data-test-id='latInput'
						value={state.lat}
						onChange={(e) =>
							setState({
								...state,
								lat: e.target.value,
							})
						}
					/>
				</Form.Group>
				<Form.Group as={Col} md='6' controlId='formLng'>
					<Form.Label>Longitude</Form.Label>
					<Form.Control
						type='text'
						placeholder='Enter longitude'
						data-test-id='lngInput'
						value={state.lng}
						onChange={(e) =>
							setState({
								...state,
								lng: e.target.value,
							})
						}
					/>
				</Form.Group>
			</Form.Row>
			<Button
				variant='primary'
				type='button'
				onClick={() => handleSubmit(state)}
			>
				Search
			</Button>
		</Form>
	);
};

export default SearchForm;
