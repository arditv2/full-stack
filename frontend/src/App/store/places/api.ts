import axios from 'axios';
import { IPlace, IApi, IApiClient, ISearchForm } from './types';
import config from '../../../config';

const instance: IApiClient = axios.create({
	baseURL: `${config.api.host}${config.api.basePath}`,
	timeout: config.api.timeout,
});

export const makeApi = (client: IApiClient): IApi => ({
	getPlaces: async (payload: ISearchForm) => {
		const { name, lat, lng } = payload;
		const query: string = `/places?provider=google&name=${encodeURI(
			name
		)}`.concat(lat && `&lat=${lat}`, lng && `&lng=${lng}`);
		const response = await client.get(query);
		return response.data as IPlace[];
	},
});

const Api: IApi = makeApi(instance);

export default Api;
