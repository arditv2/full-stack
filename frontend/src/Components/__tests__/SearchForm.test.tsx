import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { ISearchForm } from '../../App/store/places/types';
import SearchForm from '../SearchForm';

describe('SearchForm', () => {
	const handleSubmit = jest.fn();
	let wrapper: ShallowWrapper;
	const setState = jest.fn();
	const useStateSpy = jest.spyOn(React, 'useState');
	useStateSpy.mockImplementation((state) => [state, setState]);

	beforeEach(() => {
		wrapper = shallow(<SearchForm handleSubmit={handleSubmit} />);
	});

	afterEach(() => {
		jest.clearAllMocks();
	});

	it('renders nameInput FormControl', () => {
		expect(wrapper.find('FormControl[data-test-id="nameInput"]')).toHaveLength(
			1
		);
	});

	it('sets state when nameInput FormControl changes', () => {
		wrapper
			.find('FormControl[data-test-id="nameInput"]')
			.simulate('change', { target: { value: 'test-name' } });
		expect(
			wrapper.find('FormControl[data-test-id="nameInput"]').props().value
		).toEqual('test-name');
	});

	it('renders latInput FormControl', () => {
		expect(wrapper.find('FormControl[data-test-id="latInput"]')).toHaveLength(
			1
		);
	});

	it('sets state when latInput FormControl changes', () => {
		wrapper
			.find('FormControl[data-test-id="latInput"]')
			.simulate('change', { target: { value: 'test-lat' } });
		expect(
			wrapper.find('FormControl[data-test-id="latInput"]').props().value
		).toEqual('test-lat');
	});

	it('renders lngInput FormControl', () => {
		expect(wrapper.find('FormControl[data-test-id="lngInput"]')).toHaveLength(
			1
		);
	});

	it('sets state when lngInput FormControl changes', () => {
		wrapper
			.find('FormControl[data-test-id="lngInput"]')
			.simulate('change', { target: { value: 'test-lng' } });
		expect(
			wrapper.find('FormControl[data-test-id="lngInput"]').props().value
		).toEqual('test-lng');
	});

	it('renders a Button', () => {
		expect(wrapper.find('Button')).toHaveLength(1);
	});

	it('calls handleSubmit on Button click', () => {
		wrapper.find('Button').simulate('click');
		expect(handleSubmit).toHaveBeenCalled();
	});

	it('matches snapshot', () => {
		expect(wrapper).toMatchSnapshot();
	});
});
