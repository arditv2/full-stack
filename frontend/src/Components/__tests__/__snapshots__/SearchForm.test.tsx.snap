// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm matches snapshot 1`] = `
<Form
  className="mt-3 mb-3"
  inline={false}
>
  <FormGroup
    controlId="formPlaceName"
  >
    <FormLabel
      column={false}
      srOnly={false}
    >
      Place name
    </FormLabel>
    <FormControl
      data-test-id="nameInput"
      onChange={[Function]}
      placeholder="Enter place name"
      type="text"
      value=""
    />
  </FormGroup>
  <FormRow>
    <FormGroup
      as={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "displayName": "Col",
          "render": [Function],
        }
      }
      controlId="formLat"
      md="6"
    >
      <FormLabel
        column={false}
        srOnly={false}
      >
        Latitude
      </FormLabel>
      <FormControl
        data-test-id="latInput"
        onChange={[Function]}
        placeholder="Enter latitude"
        type="text"
        value=""
      />
    </FormGroup>
    <FormGroup
      as={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "displayName": "Col",
          "render": [Function],
        }
      }
      controlId="formLng"
      md="6"
    >
      <FormLabel
        column={false}
        srOnly={false}
      >
        Longitude
      </FormLabel>
      <FormControl
        data-test-id="lngInput"
        onChange={[Function]}
        placeholder="Enter longitude"
        type="text"
        value=""
      />
    </FormGroup>
  </FormRow>
  <Button
    active={false}
    disabled={false}
    onClick={[Function]}
    type="button"
    variant="primary"
  >
    Search
  </Button>
</Form>
`;
